Add lubridate

colearendt opened the pull request:
Add some `lubridate` functionality.  Several tests are failing, so we still need a bit of cleanup.

```r
> prepdata %>% mutate(as_date(fld_datetime)) %>% pull()
 [1] "2015-01-01" "2015-02-01" "2015-03-01" "2015-04-01" "2015-05-01" "2015-06-01" "2015-07-01" "2015-08-01" "2015-09-01"
[10] "2015-10-01"
> testdata %>% mutate(as_date(fld_datetime)) %>% pull()
 [1] "1-01-15"  "2-01-15"  "3-01-15"  "4-01-15"  "5-01-15"  "6-01-15"  "7-01-15"  "8-01-15"  "9-01-15"  "10-01-15"
```

Not quite sure what the best way to cause this test to succeed is... `dbtest` needs improvement for date and datetime types (`copy_to` turns them into text values, even if they are dates at the outset)

colearendt commented:
Also some precision issues with `datetime`... and `mssql` needs some serious love.  Apparently does not support `EXTRACT`, which I thought was SQL Standard:

```r
"manip(local_df) not equal to manip(remote_df) %>% integer64_fix().\n1/1 mismatches\n[1] 2018-07-19 20:12:07 - 2018-07-19 20:12:07 == -0.0274 secs"  
```

Test overview currently:
![image](https://user-images.githubusercontent.com/23075542/42968539-905d3280-8b71-11e8-84e4-97e4438114fb.png)

colearendt commented:
The test file I am currently using:

```yaml
- wday:
    mutate: wday(fld_date)
    filter: wday(fld_date)  < 5
    summarize: n_distinct(wday(fld_date))
    group_by: wday(fld_date)
    arrange: wday(fld_date)
- as_date:
    mutate: as_date(fld_datetime)
    filter: as_date(fld_datetime) > as_date("2000-01-01")
    summarize: n_distinct(as_date(fld_datetime))
    group_by: as_date(fld_datetime)
    arrange: as_date(fld_datetime)
- as_datetime:
    mutate: as_datetime(fld_datetime)
    filter: as_datetime(fld_datetime)  > as_datetime("2000-01-01")
    summarize: n_distinct(as_datetime(fld_datetime))
    group_by: as_datetime(fld_datetime)
    arrange: as_datetime(fld_datetime)
- today:
    mutate: today()
    filter: today() > as_date("2000-01-01")
    summarize: n_distinct(today())
    group_by: today()
    arrange: today() - as_date(fld_date)
- now:
    mutate: now()
    filter: now() > as_datetime("2000-01-01")
    summarize: n_distinct(now())
    group_by: now()
    arrange: now() - as_datetime(fld_datetime)
- year:
    mutate: year(fld_date)
    filter: year(fld_date)  > 2000
    summarize: n_distinct(year(fld_date))
    group_by: year(fld_date)
    arrange: year(fld_date)
- month:
    mutate: month(fld_date)
    filter: month(fld_date)  > 6
    summarize: n_distinct(month(fld_date))
    group_by: month(fld_date)
    arrange: month(fld_date)
- day:
    mutate: day(fld_date)
    filter: day(fld_date)  > 10
    summarize: n_distinct(day(fld_date))
    group_by: day(fld_date)
    arrange: day(fld_date)
- hour:
    mutate: hour(as_datetime(fld_datetime))
    filter: hour(as_datetime(fld_datetime))  > 10
    summarize: n_distinct(hour(as_datetime(fld_datetime)))
    group_by: hour(as_datetime(fld_datetime))
    arrange: hour(as_datetime(fld_datetime))
- minute:
    mutate: minute(as_datetime(fld_datetime))
    filter: minute(as_datetime(fld_datetime))  > 10
    summarize: n_distinct(minute(as_datetime(fld_datetime)))
    group_by: minute(as_datetime(fld_datetime))
    arrange: minute(as_datetime(fld_datetime))
- second:
    mutate: second(as_datetime(fld_datetime))
    filter: second(as_datetime(fld_datetime))  > 10
    summarize: n_distinct(second(as_datetime(fld_datetime)))
    group_by: second(as_datetime(fld_datetime))
    arrange: second(as_datetime(fld_datetime))

```

edgararuiz commented:
Hey! A few questions:

- What would you like specifically to have reviewed?

- Are you planning to update the PR with passing tests?

- Will this PR include compatibility with Oracle and MS SQL?

colearendt commented:
Feedback on the certain types of test failures (i.e. datetime precision, formatting, etc.) would be helpful.

I am planning to update the PR with passing tests.  Does it make sense to just start with Postgres and address Oracle / MS SQL in a separate PR?

edgararuiz commented on R/translate-sql-base.r:
```diff
@@ -175,7 +175,42 @@ base_scalar <- sql_translator(
+  day = function(x){build_sql("EXTRACT(DAY FROM ",x,")")},
+  hour = function(x){build_sql("EXTRACT(HOUR FROM ",x,")")},
+  minute = function(x){build_sql("EXTRACT(MINUTE FROM ",x,")")},
+  second = function(x){build_sql("EXTRACT(SECOND FROM ",x,")")}
```
Maybe it would be worth creating a custom function for these, especially since we'll probably use them again inside the translations for the other DB types, maybe something like this: `sql_prefix2("extract", "minute")`?

edgararuiz commented on R/translate-sql-base.r:
```diff
@@ -175,7 +175,42 @@ base_scalar <- sql_translator(
+  day = function(x){build_sql("EXTRACT(DAY FROM ",x,")")},
+  hour = function(x){build_sql("EXTRACT(HOUR FROM ",x,")")},
+  minute = function(x){build_sql("EXTRACT(MINUTE FROM ",x,")")},
+  second = function(x){build_sql("EXTRACT(SECOND FROM ",x,")")}
```
Maybe it would be worth creating a custom function for these, especially since we'll probably use them again inside the translations for the other DB types, maybe something like this: `sql_prefix2("extract", "minute")`?

hadley replied:
I think `sql_expr(extract(day %from% !!x))` isn't too bad. If you were to create a wrapper function, I think I would recommend something specifically for `extract()`.

edgararuiz commented on tests/testthat/test-translate.r:
```diff
@@ -146,6 +146,22 @@ test_that("str_trim() translates correctly ", {
+    translate_sql(
+      as_datetime("2017-01-01 01:02:03")
+    ),
+    sql("CAST('2017-01-01 01:02:03' AS DATETIME)")
```
This test fails cause the translation on line 189 in `translate_sql_base.r` uses `TIMESTAMP` and not `DATETIME`

edgararuiz commented on R/translate-sql-base.r:
```diff
@@ -175,7 +175,42 @@ base_scalar <- sql_translator(
+
+  today = function(tzone = "") {build_sql("CURRENT_DATE")},
+  now = function(tzone = "") {build_sql("CURRENT_TIMESTAMP")},
+
```
I don't think you need a curly brace for a single line funciton

hadley replied:
And I'd recommend omitting the `tzone` argument since you're not using it.

colearendt replied:
Is there any value in including the tzone argument so that there is parity with the `lubridate` function itself? I.e. makes it more clear to me that there is a disparity in the way the function is handled. 

I think my main point of difficulty here is a way of matching the package function to the implementation in dbplyr. Maybe `dbtest` can move us that direction with testing (i.e. write a failing test that shows the missing functionality).

hadley replied:
I think it's easiest to leave it off so you get an error if it's used. If you want to keep it, you'll need to explicitly throw an error, e.g.:

```R
if (!missing(tzone)) {
  stop("`tzone` not supported in SQL translation", call. = FALSE)
}
```

hadley commented:
@colearendt are you interested in finishing this off? I'll be working on dbplyr for the next few weeks.

hadley commented on R/db-postgres.r:
```diff
@@ -67,6 +67,38 @@ sql_translate_env.PostgreSQLConnection <- function(con) {
+                      , week_start = getOption("lubridate.week.start",7)
+                      , locale = Sys.getlocale("LC_TIME")) {
+        if (!label) {
+          return(build_sql(
```
Can you please eliminate the `return()`s and switch to `sql_expr()`

hadley replied:
It's useful for me to include a comment that points to the postgres docs of the function you're using (i.e. `extract()`)

hadley replied:
It might be easier to start without supporting the `label` or `abbr` arguments.

hadley commented on R/db-postgres.r:
```diff
@@ -67,6 +67,38 @@ sql_translate_env.PostgreSQLConnection <- function(con) {
         sql_expr(strpos(!!string, !!pattern) > 0L)
+      },
+      # lubridate functions
+      wday = function(x, label=FALSE, abbr = TRUE
```
Can you please apply standard tidyverse style here? e.g. https://style.tidyverse.org

hadley commented on R/translate-sql-base.r:
```diff
@@ -175,7 +175,42 @@ base_scalar <- sql_translator(
+  # SQLite - https://www.sqlite.org/lang_datefunc.html
+  #        - https://www.sqlite.org/datatype3.html
+  #
+  as_date = sql_cast("DATE"),
```
You should also include the base R `as.Date()` and `as.POSIXct()`, and Postgres also needs these equivalents.

hadley commented on tests/testthat/test-translate.r:
```diff
@@ -146,6 +146,22 @@ test_that("str_trim() translates correctly ", {
+# lubridate -------------------------------------------
 
+test_that("as_date() translates correctly ", {
+  expect_equivalent(
```
These tests should be organised to fit on a single line (or as close as possible)

hadley commented:
@edgararuiz I think that's relatively straightforward and can be done independently of this PR. I think it's just a matter of writing `escape.difftime()` and `escape.Period()`.

rlh1994 commented:
@colearendt Are you still working to provide this for mostly PostgreSQL or for other variants as well? I was planning to look at lubridate for Oracle SQL in a bit but no point in duplicating work.

colearendt commented:
@rlh1994 Definitely still working on it (and trying to cover as many variants as possible), but keep getting pulled off into other things. I am definitely not super knowledgable about Oracle SQL, so please feel free to take a stab! :smile:

colearendt commented:
Added list of not-yet-provided lubridate functions and cleaned things up. The code that generated the list:
```
ls("package:lubridate") %>% 
  purrr::map(
    ~ if (class(eval(as.name(.x))) == "function") 
      glue::glue("{.x} = sql_not_supported(\"{.x}()\"),")
  ) %>% 
  as.character() %>% 
  .[! . == "NULL"] %>%
  cat(sep = "\n")
```

colearendt commented:
I also do not understand the build failure. The `reprex.Rmd` vignette is failing on:
```
translate_sql(var = substr(x, 1, 2), con = simulate_postgres())
# Error in list2env(funs, copy_env(.parent)) : 
#  attempt to use zero-length variable name
```

This seems unrelated to anything I have changed...

hadley commented on R/backend-.R:
```diff
@@ -221,7 +222,157 @@ base_scalar <- sql_translator(
+  second = function(x) sql_expr(extract(second %from% !!x)),
+
+  `%--%` = sql_not_supported("%--%()"),
+  add_with_rollback = sql_not_supported("add_with_rollback()"),
```
Let's omit these for now. We can implement more cleanly after #207

hadley commented on R/backend-.R:
```diff
@@ -221,7 +222,157 @@ base_scalar <- sql_translator(
+
+  today = function() build_sql("CURRENT_DATE"),
+  now = function() build_sql("CURRENT_TIMESTAMP"),
+  year = function(x) sql_expr(extract(year %from% !!x)),
```
Can you please use upper case for SQL functions? i.e.

```R
sql_expr(EXTRACT(YEAR %FROM% !!x)),
```

colearendt commented:
Important to note this. It definitely is failing and causes problems using this branch... but I have no idea why:

> I also do not understand the build failure. The `reprex.Rmd` vignette is failing on:
> 
> ```
> translate_sql(substr(x, 1, 2), con = simulate_postgres())
> # Error in list2env(funs, copy_env(.parent)) : 
> #  attempt to use zero-length variable name
> ```
> This seems unrelated to anything I have changed...

hadley commented:
@colearendt oh I think it's because you did `wday <- `, instead of `wday = `. I'm finishing it off now.

colearendt commented:
😮 Woops! Sorry about that! Thanks for cleaning up!
